src/python/pants/engine/internals/graph.py

```diff
@@ -859,19 +866,33 @@ async def resolve_dependencies(
             for inference_request_type in relevant_inference_request_types
         )
 
-    # If this is a BUILD target, or no dependency inference implementation can infer dependencies on
-    # a file address's sibling files, then we inject dependencies on all the BUILD target's
-    # generated subtargets.
-    subtarget_addresses: Tuple[Address, ...] = ()
+    # If it's a target generator, inject dependencies on all of its generated targets.
+    generated_addresses: tuple[Address, ...] = ()
+    if target_types_to_generate_requests.is_generator(tgt) and not tgt.address.is_file_target:
+        generate_request = target_types_to_generate_requests[type(tgt)]
+        generated_targets = await Get(
+            GeneratedTargets, GenerateTargetsRequest, generate_request(tgt)
+        )
+        generated_addresses = tuple(t.address for t in generated_targets)
+
+    # For "file targets", inject dependencies on all siblings if that target does not have
+    # dependency inference or the implementations cannot infer dependencies on sibling files.
+    # TODO: Remove this, have the target generation rules set up the dependencies directly.
```

Comment:
That would mean that the target generation rules would be reading the config of the inference rules...? Maybe fine. In general they will be very closely related.

Reply:
Instead of trying to guess what the language author intended, they will explicitly set it. For most languages, this is based on looking if `[python-infer].imports` is set, et al. For `files` and `resources`, we never should have been inferring deps on sibling files (which will be fixed in a dedicated change because it's an API change).